\newpage
\chapter{Introduction}
\label{chapter:introduction}
\setcounter{footnote}{0}

Most protein sequences are composed from a relatively small number of
ancestral protein domain families. As our sampling of common protein
domain families has become comprehensive and deep, and with raw
sequence data continuing to grow explosively, it has become
advantageous to compare sequences against all known families rather
than all known sequences.

In this respect, protein sequence analysis is starting to become more
like speech recognition.  When you talk to your smartphone, it doesn't
compare your digitized speech to everything that's ever been said. It
compares what you say to a prebuilt dataset of statistical models of
common words and phonemes.  Using machine learning techniques, each
statistical model is trained on large datasets of examples spoken in
different accents by different speakers. Similarly, each protein
domain family typically has thousands of homologs that can be aligned
into deep multiple sequence alignments. An alignment reveals a
specific pattern of evolutionary conservation particular to that
domain's structure and function, and these patterns can be captured by
probabilistic models.

HMMER is a software package that provides tools for making statistical
models of protein and DNA sequence domain families -- called
\textbf{profile hidden Markov models}, \textbf{profile HMMs}, or just
\textbf{profiles} -- and for using these profiles to annotate new
sequences, to search sequence databases for additional homologs, and
to make deep multiple sequence alignments.  HMMER underlies several
comprehensive collections of alignments and profiles of known protein
and DNA sequence domain families, including the Pfam
database.\sidenote{\href{http://pfam.org}{pfam.org}}


\section{How to avoid reading this manual}

We hate reading documentation. If you're like us, you're thinking,
\pageref*{manualend} pages of documentation, you're joking! First I
want to know that the software compiles, runs, and gives useful
results, before I'm going to read some \pageref*{manualend} tedious
pages of someone's documentation. For fellow cynics who have seen one
too many software packages that don't work:

\begin{itemize}

\item Follow the quick installation instructions on page
  \pageref{chapter:installation}. An automated test suite is included,
  so you will know immediately if something went
  wrong.\sidenote{Nothing should go wrong.}

\item Go to the tutorial section on page \pageref{chapter:tutorial},
  which walks you through some examples of using HMMER on real data.

\end{itemize}

Everything else, you can come back and read later.



\section{Background and brief history}

A multiple sequence alignment of a homologous family of protein
domains reveals obvious patterns of site-specific evolutionary
conservation. Key residues may be highly conserved at certain
positions; some positions may tolerate certain substitutions while
conserving physiochemical properties like hydrophobicity, charge, or
size; some positions may be evolutionarily near-neutral and variable;
insertions and deletions are tolerated at some positions better than
others.  A \textbf{profile} is a position-specific scoring model that
describes which symbols are likely to be observed and how frequently
insertions/deletions occur at each position (column) of a multiple
sequence alignment.

Pairwise sequence alignment methods such as BLAST and FASTA use
position-\emph{independent} subsitution score matrices such as BLOSUM
and PAM, but the desirability of position-\emph{specific} models was
recognized even before BLAST and FASTA were written.\cite{Doolittle81}
Several groups introduced different position-specific alignment
scoring approaches in the 1980's. The name ``profiles'', introduced by
Gribskov and colleagues,\cite{Gribskov87} was a name that stuck.

Profiles have a lot of parameters.  The BLOSUM and PAM amino acid
substitution matrices have only 210 free parameters (20 $\times$ 20,
symmetric), and those parameters are averages over large collections
of many different known sequence alignments. A profile typically has
at least 22 parameters\marginnote{There's $\sim$22 parameters per
  position because there's 20 residue scores, plus gap-open and
  gap-extend penalties for starting or extending an insertion or
  deletion.}  for each of the $\sim$200 consensus positions or so in a
typical protein domain, and these thousands of parameters are
estimated for one particular sequence family alignment, not averaged
across many. Early profile methods were vexed by a lack of theoretical
guidance on how to parameterize these models effectively, especially
for insertion and deletions.

In the 1990's, Anders Krogh, David Haussler, and co-workers at UC
Santa Cruz recognized a parallel between profiles and widely used
speech recognition techniques, and they introduced \textbf{profile
  hidden Markov models (profile HMMs)}.\cite{Krogh94} HMMs had been
used in biology before, but the Krogh paper had dramatic impact
because HMM technology was so well suited to addressing the vexing
parameterization problem. HMMs have a formal probabilistic basis,
allowing us to use probability theory to set and to interpret the
large number of free parameters in a profile, including the
position-specific gap and insertion parameters. The methods are
mathematically consistent and therefore automatable, which was crucial
in allowing people to make libraries of hundreds of profile HMMs and
apply them on a large scale to whole genome analysis.  One such
database of protein domain models is Pfam.\cite{Sonnhammer97}
Historically, Pfam and HMMER have been developed in parallel.

The first implementations of profile HMMs were computationally
intensive, including HMMER1 (1995) and HMMER2 (1998), but HMMER3 is
now typically faster than BLASTP or FASTA searches even though it uses
more complex models.\marginnote{For DNA searches, BLASTN remains about
  two orders of magnitude faster than HMMER DNA searches, but is less
  sensitive.}

\section{Problems HMMER is designed to address}

\paragraph{Sensitive homology searches.}
You're working on a specific sequence family, and you have carefully
constructed a representative multiple sequence alignment. The HMMER
\mono{hmmbuild} program lets you build a profile from your alignment,
and the \mono{hmmsearch} program lets you search your profile against
a sequence database looking for homologs.

\paragraph{... even for single sequence queries.}
HMMER3 also works for single sequence comparisons, not just for
multiple alignments. Pairwise sequence comparison is just a special
case of profile HMMs. HMMER can use a BLOSUM substitution matrix to
parameterize a profile built from just one sequence.  HMMER3 includes
two programs for searching protein databases with single query
sequences: \mono{phmmer} and \mono{jackhmmer}. We believe
\mono{phmmer} is superior in many respects to BLASTP, and
\mono{jackhmmer} to PSI-BLAST.

\paragraph{Automated annotation of protein domains.}
Various large databases of curated alignments and HMMER models of
known domains are available, including Pfam and others.  Many ``top
ten protein domains'' lists, a \emph{de rigueur} table in genome
analysis papers, depend on HMMER-based annotation. HMMER3's
\mono{hmmscan} program lets you scan a sequence against a profile
database to parse the sequence into its component domains.

\paragraph{Curated collections of deep multiple alignments.}  There are thousands of
sequence families, some of which comprise hundreds of thousands of
sequences, and the raw sequence databases continue to double every
year or so. Clustering the entire sequence database into family
alignments is a hopeless task for manual curation, but some sort of
manual curation remains necessary for high-quality, biologically
relevant multiple alignments. Databases like Pfam are constructed by
distinguishing between a stable curated ``seed'' alignment of a small
number of representative sequences, and ``full'' alignments of all
detectable homologs. HMMER is used to make a model of the seed and to
search the database for homologs, and the \mono{hmmalign} program can
automatically produce the full alignment by aligning every sequence to
the seed consensus. \mono{hmmalign} scales to alignments of 
millions of sequences.



\section{Ensemble algorithms, \emph{not} optimal alignment}

Most sequence search tools look for optimal high-scoring
alignments. However, sequence alignments are uncertain, and the more
distantly related sequences are, the more uncertain their alignment
is. Instead of using optimal alignment algorithms, HMMER uses ensemble
algorithms that consider all possible alignments, weighted by their
relative likelihood.\marginnote{In HMM jargon, we use the Forward
  algorithm (and variants of it), not the Viterbi algorithm.} This is
one reason that HMMER gets more power than tools that depend on single
optimal alignment.

The use of ensemble algorithms shows up in several features that
differentiate HMMER from some other methods:

\paragraph{Explicit representation of alignment uncertainty.}
  When HMMER shows an alignment, it also calculates how much
  probability mass that this alignment has in the ensemble -- which
  means HMMER can annotate a probabilistic confidence in an alignment,
  or in each individual aligned residue. Some downstream analyses that
  depend on alignments (such as phylogenetic tree inference) benefit
  from being able to distinguish confidently aligned residues.

\paragraph{Sequence scores, not alignment scores.} HMMER's
  log-odds scores for a sequence aren't optimal alignment scores; they
  are summed over the posterior alignment ensemble. Statistical
  inference theory says that scores based on a single optimal
  alignment are an approximation that breaks down when alignments are
  uncertain. HMMER's calculation is the full, unapproximated
  calculation.
 
\paragraph{Different speed heuristics.} The ensemble (Forward) algorithm is more
  computationally intensive than optimal alignment algorithms.
  HMMER3's acceleration strategy is quite different from
  BLAST's.\cite{Eddy11}.  HMMER implements heuristic accelerations of
  the HMM Forward algorithm using vectorization technology available
  on modern processors.\vspace{1em}

Individually, none of these points is new. As far as alignment
ensembles go, one reason why hidden Markov models were so
theoretically attractive in the first place for sequence analysis is
that they are good probabilistic models for explicitly dealing with
alignment uncertainty. The SAM profile HMM software from UC Santa Cruz
has always used full probabilistic inference (the HMM Forward/Backward
algorithms) as opposed to optimal alignment scores (the HMM Viterbi
algorithm). HMMER2 had the full HMM inference algorithms available as
command-line options, but it used Viterbi optimal alignment by
default, in part for speed reasons.

One reason why it's been hard to deploy sequence scores for practical
large-scale use is that we haven't known how to accurately calculate
the statistical significance of a log-odds score that's been summed
over alignment uncertainty. Accurate statistical significance
estimates are essential when one is trying to discriminate homologs
from millions of unrelated sequences in a large sequence database
search. The statistical significance of optimal local alignment scores
is calculated by Karlin/Altschul statistics.\cite{Karlin90}
Karlin/Altschul statistics are one of the most important and
fundamental advances introduced by BLAST.  However, Karlin/Altschul
theory \emph{doesn't} apply to HMMER's ensemble log-odds sequence
scores (HMM ``Forward scores''). The statistical significance
(E-values, or expectation values) of HMMER sequence scores is
determined by using a theoretical conjecture about the statistical
properties of ensemble log-odds scores which have been supported by
numerical simulation experiments.\cite{Eddy08}

And as far as speed goes, the pioneers of heuristic acceleration of
sequence database searches are the folks behind BLAST and FASTA, who
developed effective heuristics that closely approximate an
unaccelerated Smith/Waterman dynamic programming search.  The first
implementations of profile HMM methods used dynamic programming
without heuristics (the profile HMM Viterbi algorithm is essentially
Smith/Waterman, just with position-specific probability scores), so
they were more comparable in speed to Smith/Waterman than to BLAST.
Using the Forward algorithm slowed them down still more. It was a
while before we invested the time to develop heuristic acceleration of
profile HMM methods. A principal design goal in HMMER3 was to achieve
at least rough speed parity with BLAST and FASTA.


\section{Assumptions and limitations of profile HMMs}

Profile HMMs are primary sequence consensus models.  They assume that
the residue at a particular position is independent of the residues at
all other positions, and they neglect any higher-order correlations.
Profile HMMs are often not good models of structural RNAs, for
instance, because an HMM is not an adequate model of correlated base
pairs.\marginnote{Our Infernal software provides better tools for
  structural RNA sequence analysis, using \textbf{profile stochastic
    context-free grammars}, a more complex class of probability model
  than HMMs.}

A profile HMM also lacks any explicit model of the phylogenetic
relationships among a set of homologous sequences. Sequences are
instead assumed to be independently generated from the profile, which
is tantamount to assuming a star phylogeny with fixed branch
lengths. Ad hoc sequence weighting techniques are used to compensate
for the fact that typical multiple alignments include many redundant,
closely related sequences.


\section{What's still missing}

Two of the more important holes for us are:

\paragraph{Translated comparisons.} We'd love to have the HMM
equivalents of BLASTX, TBLASTN, and TBLASTX. They'll come. In the
meantime, we translate DNA sequence to six frames, and search
hypothetical ORFs as protein sequences.

\paragraph{Profile/profile comparison.} A number of pioneering papers and
software packages have demonstrated the power of profile/profile
comparison for even more sensitive remote homology detection. We will
aim to develop profile HMM methods in HMMER with improved detection
power, and at HMMER3 speed. In the meantime, check out
HHBLITS from Johannes S\"oding's
group.\sidenote{\href{https://toolkit.tuebingen.mpg.de/\#/tools/hhblits}{toolkit.tuebingen.mpg.de/\#/tools/hhblits}}


\section{How to learn more}

Our book \emph{Biological Sequence Analysis}\cite{Durbin98} describes the basic
theory behind profile HMMs and HMMER.

HMMER's open source development code is available on
GitHub.\sidenote{\href{http://github.com/EddyRivasLab/hmmer}{github.com/EddyRivasLab/hmmer}}
The GitHub issue tracker is the best way to give us suggestions,
feature requests, bug reports, and pull requests.

\textbf{Cryptogenomicon}\sidenote{\href{http://cryptogenomicon.org/}{cryptogenomicon.org}}
  is a blog where I sometimes talk about issues as they arise in HMMER, and
where you can comment or follow the discussion.


\section{How to avoid using this software (links to similar software)}

Other implementations of profile HMM methods and position-specific
scoring matrix methods are available, including:

\begin{center}
\begin{tabular}{lp{5in}l}
Software  &   URL \\ \hline
HH-SUITE  & \href{http://www.soeding.genzentrum.lmu.de/software-and-servers-2/}{www.soeding.genzentrum.lmu.de/software-and-servers-2}\\
PSI-BLAST & \href{https://blast.ncbi.nlm.nih.gov/}{blast.ncbi.nlm.nih.gov}\\
PFTOOLS   & \href{http://web.expasy.org/pftools/}{web.expasy.org/pftools}\\
SAM       & \href{https://compbio.soe.ucsc.edu/sam.html}{compbio.soe.ucsc.edu/sam.html}\\
\end{tabular}
\end{center}














  









